streams/src/test/java/org/apache/kafka/streams/integration/GlobalStateReprocessTest.java

```diff
@@ -0,0 +1,205 @@
+        streamsConfiguration.put(StreamsConfig.STATESTORE_CACHE_MAX_BYTES_CONFIG, 0);
+        streamsConfiguration.put(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG, 100L);
+
+        final KeyValueStoreBuilder<String, Long> storeBuilder = new KeyValueStoreBuilder<>(
+            Stores.persistentKeyValueStore(globalStore),
+            Serdes.String(),
+            Serdes.Long(),
+            mockTime);
+
+        final ProcessorSupplier<String, Long, Void, Void> processorSupplier;
+        processorSupplier = () -> new ContextualProcessor<String, Long, Void, Void>() {
+            @Override
+            public void process(final Record<String, Long> record) {
+                final KeyValueStore<String, Long> stateStore =
+                    context().getStateStore(storeBuilder.name());
+                stateStore.put(
+                    record.key() + "- this is the right value.",
+                    record.value()
+                );
+            }
+        };
+
+        builder.addGlobalStore(
+            storeBuilder,
+            globalStoreTopic,
+            Consumed.with(Serdes.String(), Serdes.Long()),
+            processorSupplier
+        );
+    }
+
+    @AfterEach
+    public void after() throws Exception {
+        if (kafkaStreams != null) {
+            kafkaStreams.close();
+        }
+        IntegrationTestUtils.purgeLocalStreamsState(streamsConfiguration);
+    }
+
+    @Test
+    public void shouldReprocessWithUserProvidedStore() throws Exception {
+        kafkaStreams = new KafkaStreams(builder.build(), streamsConfiguration);
+        populateTopics(globalStoreTopic);
+
+        kafkaStreams.start();
+
+        TestUtils.waitForCondition(
+            () -> !storeContents(kafkaStreams).isEmpty(),
+            30000,
+            "Has not processed record within 30 seconds");
+
+        assertThat(storeContents(kafkaStreams).get(0), containsString("- this is the right value."));
+
+
+        kafkaStreams.close();
+        kafkaStreams.cleanUp();
+
+        kafkaStreams = new KafkaStreams(builder.build(), streamsConfiguration);
+        kafkaStreams.start();
+
+        TestUtils.waitForCondition(
+            () -> !storeContents(kafkaStreams).isEmpty(),
+            30000,
+            "Has not processed record within 30 seconds");
+
+        assertThat(storeContents(kafkaStreams).get(0), containsString("- this is the right value."));
+    }
+
+    private void createTopics() throws Exception {
+        globalStoreTopic = "global-store-topic";
+        CLUSTER.createTopic(globalStoreTopic);
+    }
+
+    private void populateTopics(final String topicName) throws Exception {
+        IntegrationTestUtils.produceKeyValuesSynchronously(
+            topicName,
+            Arrays.asList(
+                new KeyValue<>("A", 1L),
+                new KeyValue<>("B", 2L),
+                new KeyValue<>("C", 3L),
+                new KeyValue<>("D", 4L)),
```

Comment:
Why are we writing 4 records? Seems one would be sufficient given what assertions the test makes?

Reply:
There was a couple times it made testing easier. We can move them down to one if you would prefer